### PR TITLE
NXP-27010: handle different REST API versions

### PR DESCRIPTION
**Pull request by troger**
WIP / To review

### CODE REVIEW

**Comment by troger** [comment]
TP: https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-troger/342/

**Comment by troger** [comment]
T&P: https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-troger/346/

**Comment by troger** [comment]
T&P: https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-troger/347/

File: nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/APIVersion.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    /**
+     * Returns the {@code APIVersion} object for the given {@code version}.
+     *
+     * @throws NuxeoException if the {@code version} is not a valid REST API version.
+     */
+    public static APIVersion of(int version) {
+        APIVersion apiVersion = VALID_VERSIONS.get(version);
+        if (apiVersion == null) {
+            throw new NuxeoException(
+                    String.format("%s is not part of the valid versions: %s", version, VALID_VERSIONS.keySet()),
```
Comment (kevinleturc) [inline nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/APIVersion.java]: Do we want versions sorted?

File: nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/APIVersion.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        return LATEST_VERSION;
+    }
+
+    protected final int version;
+
+    APIVersion(int version) {
+        this.version = version;
+    }
+
+    public boolean eq(APIVersion other) {
```
Comment (kevinleturc) [inline nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/APIVersion.java]: What about adding `@NotNull` for all parameters? Just to help IDEs.

File: nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/registry/context/RenderingContextImpl.java [hunk 3]
```diff
@@ -331,6 +333,12 @@ public void addParameterListValues(String name, List<?> values) {
         parameters.computeIfAbsent(realName, key -> new CopyOnWriteArrayList<>()).addAll(values);
     }
 
+    @Override
+    public APIVersion getAPIVersion() {
+        APIVersion apiVersion = getParameter(API_VERSION_ATTRIBUTE_NAME);
+        return apiVersion != null ? apiVersion : APIVersion.latest();
```
Comment (kevinleturc) [inline nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/registry/context/RenderingContextImpl.java]: What do you think about:
```suggestion
        return Objects.requireNonNullElseGet(apiVersion, APIVersion::latest);
```

File: nuxeo-core/nuxeo-core-io/src/test/java/org/nuxeo/ecm/core/io/APIVersionTest.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * @since 11.1
+ */
+public class APIVersionTest {
+
+    @Test
+    public void testLatestVersion() {
+        APIVersion expectedLatest = APIVersion.VALID_VERSIONS.values()
+                                                           .stream()
```
Comment (kevinleturc) [inline nuxeo-core/nuxeo-core-io/src/test/java/org/nuxeo/ecm/core/io/APIVersionTest.java]: Formatting.

**Comment by kevinleturc** [comment]
👍

**Comment by kevinleturc** [comment]
👍

**Pull request merged**